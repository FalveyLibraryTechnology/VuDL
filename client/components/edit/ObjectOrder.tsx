import React, { useState } from "react";
import { useEditorContext } from "../../context/EditorContext";
import { getObjectDirectChildPidsUrl, getObjectSortOnUrl, getPositionInParentUrl } from "../../util/routes";
import { useFetchContext } from "../../context/FetchContext";

export interface ObjectOrderProps {
    pid: string;
}

const ObjectOrder = ({ pid }: ObjectOrderProps): React.ReactElement => {
    const {
        state: { objectDetailsStorage },
        action: { clearPidFromChildListStorage, removeFromObjectDetailsStorage },
    } = useEditorContext();
    const {
        action: { fetchJSON, fetchText },
    } = useFetchContext();
    const [statusMessage, setStatusMessage] = useState<string>("");
    const currentSort = objectDetailsStorage?.[pid]?.sortOn ?? "title";
    const childPageSize = 1000;
    const solrSort = "title_sort ASC,id ASC";
    const changeSort = async (sort: string) => {
        if (!confirm("Are you sure you want to change the sort?")) {
            return;
        }
        setStatusMessage("Saving...");
        let offset = 0;
        let response = null;
        let result = "";
        let currentStatus = "";
        currentStatus = `Changing ${pid} sort to ${sort}`;
        setStatusMessage(currentStatus);
        const sortResult = await fetchText(getObjectSortOnUrl(pid), { method: "PUT", body: sort });
        if (sortResult != "ok") {
            setStatusMessage(`${currentStatus} -- unexpected error`);
            return;
        }
        let errorOccurred = false;
        do {
            const url = getObjectDirectChildPidsUrl(pid, offset, childPageSize, solrSort);
            response = await fetchJSON(url);
            for (const x in response.docs ?? []) {
                const targetPid = response.docs[x].id;
                const positionUrl = getPositionInParentUrl(targetPid, pid);
                if (sort === "custom") {
                    const newPosition = offset + parseInt(x) + 1;
                    currentStatus = `Setting ${targetPid} to position ${newPosition}`;
                    setStatusMessage(currentStatus);
                    result = await fetchText(positionUrl, { method: "PUT", body: newPosition });
                } else {
                    currentStatus = `Clearing order for ${targetPid}`;
                    setStatusMessage(currentStatus);
                    result = await fetchText(positionUrl, { method: "DELETE" });
                }
                if (result != "ok") {
                    setStatusMessage(`${currentStatus} -- unexpected error`);
                    errorOccurred = true;
                    break;
                }
                removeFromObjectDetailsStorage(targetPid);
            }
            offset += childPageSize;
        } while (!errorOccurred && offset < (response.numFound ?? 0));
        if (!errorOccurred) {
            setStatusMessage("");
        }
        removeFromObjectDetailsStorage(pid);
        clearPidFromChildListStorage(pid);
    };
    const otherSort = currentSort === "title" ? "custom" : "title";
    return statusMessage.length > 0 ? (
        <div>{statusMessage}</div>
    ) : (
        <div>
            Current sort: {currentSort}
            <button
                onClick={() => {
                    changeSort(otherSort);
                }}
            >
                Change to {otherSort}
            </button>
        </div>
    );
};

export default ObjectOrder;
