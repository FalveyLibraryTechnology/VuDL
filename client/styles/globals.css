:root,
[color-scheme="light"] {
    --body-bg: #fafbfc;
    --heading-color: #4d4d4d;
    --link-color: #008aed;

    --btn-bg: #485a6c;
    --btn-text: #ffffff;
    --btn-border: #275575;
    --btn-focus-bg: #3498db;

    --btn-primary-bg: #27ae60;
    --btn-primary-text: #ffffff;
    --btn-primary-border: #275575;
    --btn-primary-focus-bg: #155724;

    --btn-danger-bg: red;
    --btn-danger-text: #ffffff;
    --btn-danger-border: #a10705;
    --btn-danger-focus-bg: #a10705;

    --btn-logout-bg: #1abc9c;
    --btn-logout-text: #ffffff;
    --btn-logout-border: #16a085;
    --btn-logout-focus-bg: #1abc9c;

    --btn-toggle-bg: #7f8c8d;
    --btn-toggle-text: #ffffff;
    --btn-toggle-border: #34495e;
    --btn-toggle-focus-bg: #53626f;

    --group-border: #53626f;

    --input-border: #6a737d;
    --input-focus-border: #2188ff;

    --label-hover: #c8e1ff;

    --page-list-bg: #ffffff;
    --page-list-border: #c1c1c1;

    --thumbnail-bg: #e1e1e1;
    --thumbnail-focus-bg: #51aded;
    --thumbnail-number-bg: #ffffff;
    --thumbnail-number-text: #000000;
    --thumbnail-preview-border: #000000;

    --tree-item-focus-bg: #000000;
    --tree-item-focus-text: #ffffff;

    --shadow-gray: #c8e1ff;
}

[color-scheme="dark"] {
}
