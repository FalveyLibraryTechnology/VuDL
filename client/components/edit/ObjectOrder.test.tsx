import React from "react";
import { beforeEach, describe, expect, it, jest } from "@jest/globals";
import { shallow, mount } from "enzyme";
import { act } from "react-dom/test-utils";
import { waitFor } from "@testing-library/react";
import toJson from "enzyme-to-json";
import ObjectOrder from "./ObjectOrder";

const mockUseEditorContext = jest.fn();
jest.mock("../../context/EditorContext", () => ({
    useEditorContext: () => {
        return mockUseEditorContext();
    },
}));

const mockUseFetchContext = jest.fn();
jest.mock("../../context/FetchContext", () => ({
    useFetchContext: () => {
        return mockUseFetchContext();
    },
}));

describe("ObjectOrder", () => {
    let editorValues;
    let fetchContextValues;
    const pid = "foo:123";
    beforeEach(() => {
        editorValues = {
            state: {
                objectDetailsStorage: {},
            },
            action: {
                clearPidFromChildListStorage: jest.fn(),
                removeFromObjectDetailsStorage: jest.fn(),
            },
        };
        mockUseEditorContext.mockReturnValue(editorValues);
        fetchContextValues = {
            action: {
                fetchJSON: jest.fn(),
                fetchText: jest.fn(),
            },
        };
        mockUseFetchContext.mockReturnValue(fetchContextValues);
    });

    afterEach(() => {
        jest.resetAllMocks();
    });

    it("defaults to title order for a pending object", () => {
        const wrapper = shallow(<ObjectOrder pid={pid} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("displays a custom-sorted object correctly", () => {
        editorValues.state.objectDetailsStorage[pid] = { pid, sortOn: "custom" };
        const wrapper = shallow(<ObjectOrder pid={pid} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("can be aborted via confirmation dialog", async () => {
        editorValues.state.objectDetailsStorage[pid] = { pid, sortOn: "custom" };
        fetchContextValues.action.fetchJSON.mockResolvedValue({ numFound: 0 });
        fetchContextValues.action.fetchText.mockResolvedValue("ok");
        const confirmSpy = jest.spyOn(window, "confirm").mockReturnValue(false);
        let wrapper;
        await act(async () => {
            wrapper = mount(<ObjectOrder pid={pid} />);
        });
        await act(async () => {
            wrapper.find("button").simulate("click");
        });
        expect(fetchContextValues.action.fetchText).not.toHaveBeenCalled();
        expect(confirmSpy).toHaveBeenCalledTimes(1);
        expect(editorValues.action.removeFromObjectDetailsStorage).not.toHaveBeenCalled();
        expect(editorValues.action.clearPidFromChildListStorage).not.toHaveBeenCalled();
    });

    it("saves changes to an empty collection correctly", async () => {
        editorValues.state.objectDetailsStorage[pid] = { pid, sortOn: "custom" };
        fetchContextValues.action.fetchJSON.mockResolvedValue({ numFound: 0 });
        fetchContextValues.action.fetchText.mockResolvedValue("ok");
        const confirmSpy = jest.spyOn(window, "confirm").mockReturnValue(true);
        let wrapper;
        await act(async () => {
            wrapper = mount(<ObjectOrder pid={pid} />);
        });
        await act(async () => {
            wrapper.find("button").simulate("click");
        });
        await waitFor(() =>
            expect(fetchContextValues.action.fetchText).toHaveBeenCalledWith(
                "http://localhost:9000/api/edit/object/foo%3A123/sortOn",
                { body: "title", method: "PUT" },
            ),
        );
        expect(confirmSpy).toHaveBeenCalledTimes(1);
        expect(editorValues.action.removeFromObjectDetailsStorage).toHaveBeenCalledWith(pid);
        expect(editorValues.action.clearPidFromChildListStorage).toHaveBeenCalledWith(pid);
    });

    it("handles sort save failure", async () => {
        editorValues.state.objectDetailsStorage[pid] = { pid, sortOn: "custom" };
        fetchContextValues.action.fetchJSON.mockResolvedValue({ numFound: 0 });
        fetchContextValues.action.fetchText.mockResolvedValue("kaboom");
        const confirmSpy = jest.spyOn(window, "confirm").mockReturnValue(true);
        let wrapper;
        await act(async () => {
            wrapper = mount(<ObjectOrder pid={pid} />);
        });
        await act(async () => {
            wrapper.find("button").simulate("click");
        });
        await waitFor(() =>
            expect(fetchContextValues.action.fetchText).toHaveBeenCalledWith(
                "http://localhost:9000/api/edit/object/foo%3A123/sortOn",
                { body: "title", method: "PUT" },
            ),
        );
        expect(confirmSpy).toHaveBeenCalledTimes(1);
        wrapper.update();
        expect(wrapper.text()).toEqual("Changing foo:123 sort to title -- unexpected error");
        expect(editorValues.action.removeFromObjectDetailsStorage).not.toHaveBeenCalled();
        expect(editorValues.action.clearPidFromChildListStorage).not.toHaveBeenCalled();
    });

    it("handles child save failure correctly when switching to custom", async () => {
        editorValues.state.objectDetailsStorage[pid] = { pid, sortOn: "title" };
        fetchContextValues.action.fetchJSON.mockResolvedValue({ numFound: 1, docs: [{ id: "foo:125" }] });
        fetchContextValues.action.fetchText.mockResolvedValueOnce("ok");
        fetchContextValues.action.fetchText.mockResolvedValueOnce("not ok");
        const confirmSpy = jest.spyOn(window, "confirm").mockReturnValue(true);
        let wrapper;
        await act(async () => {
            wrapper = mount(<ObjectOrder pid={pid} />);
        });
        await act(async () => {
            wrapper.find("button").simulate("click");
        });
        await waitFor(() =>
            expect(fetchContextValues.action.fetchText).toHaveBeenCalledWith(
                "http://localhost:9000/api/edit/object/foo%3A125/positionInParent/foo%3A123",
                { body: 1, method: "PUT" },
            ),
        );
        expect(confirmSpy).toHaveBeenCalledTimes(1);
        expect(editorValues.action.removeFromObjectDetailsStorage).toHaveBeenCalledWith(pid);
        expect(editorValues.action.clearPidFromChildListStorage).toHaveBeenCalledWith(pid);
        wrapper.update();
        expect(wrapper.text()).toEqual("Setting foo:125 to position 1 -- unexpected error");
    });

    it("handles child save failure correctly when switching to title", async () => {
        editorValues.state.objectDetailsStorage[pid] = { pid, sortOn: "custom" };
        fetchContextValues.action.fetchJSON.mockResolvedValue({ numFound: 1, docs: [{ id: "foo:125" }] });
        fetchContextValues.action.fetchText.mockResolvedValueOnce("ok");
        fetchContextValues.action.fetchText.mockResolvedValueOnce("not ok");
        const confirmSpy = jest.spyOn(window, "confirm").mockReturnValue(true);
        let wrapper;
        await act(async () => {
            wrapper = mount(<ObjectOrder pid={pid} />);
        });
        await act(async () => {
            wrapper.find("button").simulate("click");
        });
        await waitFor(() =>
            expect(fetchContextValues.action.fetchText).toHaveBeenCalledWith(
                "http://localhost:9000/api/edit/object/foo%3A125/positionInParent/foo%3A123",
                { method: "DELETE" },
            ),
        );
        expect(confirmSpy).toHaveBeenCalledTimes(1);
        expect(editorValues.action.removeFromObjectDetailsStorage).toHaveBeenCalledWith(pid);
        expect(editorValues.action.clearPidFromChildListStorage).toHaveBeenCalledWith(pid);
        wrapper.update();
        expect(wrapper.text()).toEqual("Clearing order for foo:125 -- unexpected error");
    });

    it("successfully updates children", async () => {
        editorValues.state.objectDetailsStorage[pid] = { pid, sortOn: "custom" };
        fetchContextValues.action.fetchJSON.mockResolvedValue({ numFound: 1, docs: [{ id: "foo:125" }] });
        fetchContextValues.action.fetchText.mockResolvedValue("ok");
        const confirmSpy = jest.spyOn(window, "confirm").mockReturnValue(true);
        let wrapper;
        await act(async () => {
            wrapper = mount(<ObjectOrder pid={pid} />);
        });
        await act(async () => {
            wrapper.find("button").simulate("click");
        });
        await waitFor(() =>
            expect(fetchContextValues.action.fetchText).toHaveBeenCalledWith(
                "http://localhost:9000/api/edit/object/foo%3A125/positionInParent/foo%3A123",
                { method: "DELETE" },
            ),
        );
        expect(confirmSpy).toHaveBeenCalledTimes(1);
        expect(editorValues.action.removeFromObjectDetailsStorage).toHaveBeenCalledWith(pid);
        expect(editorValues.action.clearPidFromChildListStorage).toHaveBeenCalledWith(pid);
        expect(fetchContextValues.action.fetchText).toHaveBeenCalledWith(
            "http://localhost:9000/api/edit/object/foo%3A123/sortOn",
            { body: "title", method: "PUT" },
        );
        expect(fetchContextValues.action.fetchText).toHaveBeenCalledTimes(2);
    });
});
