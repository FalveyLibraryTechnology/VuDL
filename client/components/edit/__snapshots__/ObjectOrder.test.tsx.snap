// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectOrder defaults to title order for a pending object 1`] = `
<div>
  Current sort: 
  title
  <button
    onClick={[Function]}
  >
    Change to 
    custom
  </button>
</div>
`;

exports[`ObjectOrder displays a custom-sorted object correctly 1`] = `
<div>
  Current sort: 
  custom
  <button
    onClick={[Function]}
  >
    Change to 
    title
  </button>
</div>
`;
